src/passes/OptimizeInstructions.cpp

```diff
@@ -1605,6 +1605,49 @@ struct OptimizeInstructions
       curr->op = Abstract::getBinary(type, Ne);
       return curr;
     }
+    // (unsigned)x <= -1   ==>   1
+    if (matches(curr, binary(LeU, pure(&left), ival(-1)))) {
+      right->value = Literal::makeOne(Type::i32);
+      right->type = Type::i32;
+      return right;
+    }
+    {
+      // (signed)x < (i32|i64).max_s   ==>   x != (i32|i64).max_s
+      Const* c;
+      if (matches(curr, binary(LtS, any(), ival(&c))) &&
+          c->value.isSignedMax()) {
+        curr->op = Abstract::getBinary(type, Ne);
+        return curr;
+      }
+      // (signed)x <= (i32|i64).max_s   ==>   i32(1)
+      if (matches(curr, binary(LeS, pure(&left), ival(&c))) &&
+          c->value.isSignedMax()) {
+        right->value = Literal::makeOne(Type::i32);
+        right->type = Type::i32;
+        return right;
+      }
+      // (signed)x >= (i32|i64).min_s   ==>   i32(1)
+      if (matches(curr, binary(GeS, pure(&left), ival(&c))) &&
+          c->value.isSignedMin()) {
+        right->value = Literal::makeOne(Type::i32);
+        right->type = Type::i32;
+        return right;
+      }
+      // (signed)x < (i32|i64).min_s   ==>   i32(0)
+      if (matches(curr, binary(LtS, pure(&left), ival(&c))) &&
+          c->value.isSignedMin()) {
+        right->value = Literal::makeZero(Type::i32);
+        right->type = Type::i32;
+        return right;
+      }
+      // (signed)x > (i32|i64).max_s   ==>   i32(0)
+      if (matches(curr, binary(GtS, pure(&left), ival(&c))) &&
+          c->value.isSignedMax()) {
+        right->value = Literal::makeZero(Type::i32);
+        right->type = Type::i32;
+        return right;
+      }
+    }
```

Comment:
It looks like there could be more cases handled: `x >= max -> x == max`, x <= min -> x == min`. It would also be good to order these by operation and constant. Right now I can't tell whether there's any ordering scheme.

Reply:
I guess should be complete now

Reply:
Do we also have `(unsigned)x >= 0 -> 1` somewhere?

Reply:
nope. I could add it in this PR

Reply:
Done

Reply:
Nice! This looks like the complete set of transformations now. My only remaining qualm is that the unsigned transformations are ordered differently from the signed transformations (ordered by input constant vs ordered by output pattern). If you could standardize the ordering scheme across both, that would be great. After that I think this will be good to go.

Reply:
Ok, I'll do reordering after refuzzing)

Reply:
Btw I prefer do transform which folded to const first due to it will cause to early exit from opt loop.

Reply:
Alright. I split rules on two groups. One is folded to constant and other just switch op. Both ordered by binary operator ascending.

Reply:
Great, thanks!